### PR TITLE
Cleanup `ChannelKeyManager`

### PR DESCRIPTION
**Pull request by pm47**
There was a confusion between `fundingKeyPath` and `channelKeyPath`.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala [hunk 1]
```diff
@@ -38,8 +38,23 @@ object LocalChannelKeyManager {
...
+ * The resulting paths looks like so on mainnet:
+ * {{{
+ *  funding:
+ *   - initial/static
+ *     47' / 1' / <fundingKeyPath> / <1 or 0> / 0
+ *   - dynamic
+ *     47' / 1' / <fundingKeyPath> / <1 or 0> / 6 / <index>
+ *
+ *  all others (payment, revocation, htlc, etc.):
+ *     47' / 1' / <channelKeyPath> / <1-5> / <index>
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala:56]: Please double check this as it is easy to miss something.

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala [hunk 2]
```diff
@@ -58,17 +73,7 @@ class LocalChannelKeyManager(seed: ByteVector, chainHash: ByteVector32) extends
...
-
-  private def fundingPrivateKey(channelKeyPath: DeterministicWallet.KeyPath): ExtendedPrivateKey = privateKeys.get(internalKeyPath(channelKeyPath, hardened(0)))
-
-  private def revocationSecret(channelKeyPath: DeterministicWallet.KeyPath): ExtendedPrivateKey = privateKeys.get(internalKeyPath(channelKeyPath, hardened(1)))
-
-  private def paymentSecret(channelKeyPath: DeterministicWallet.KeyPath): ExtendedPrivateKey = privateKeys.get(internalKeyPath(channelKeyPath, hardened(2)))
-
-  private def delayedPaymentSecret(channelKeyPath: DeterministicWallet.KeyPath): ExtendedPrivateKey = privateKeys.get(internalKeyPath(channelKeyPath, hardened(3)))
-
-  private def htlcSecret(channelKeyPath: DeterministicWallet.KeyPath): ExtendedPrivateKey = privateKeys.get(internalKeyPath(channelKeyPath, hardened(4)))
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala:71]: I believe those can be removed now @sstone?

File: eclair-core/src/main/scala/fr/acinq/eclair/json/JsonSerializers.scala [hunk 3]
```diff
@@ -216,8 +216,9 @@ object TransactionSerializer extends MinimalSerializer({
   ))
 })
 
+
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/json/JsonSerializers.scala:219]: nit: redundant line

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala (hunk 1 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala:56]: > That also means we don't have to use 6 for the splice funding txs, if <1-5> are used with a different path anyway

If we use `this/path` for the initial funding tx, and `this/path/<index>` for the subsequent funding txs, it is my understanding that leaking the private key for the initial funding key will compromise all the child keys. So I think we do in fact need a different path.

> it's hard to figure out which paths are already used

Indeed, that's why I inlined the values in my `lightning-kmp` rework: https://github.com/ACINQ/lightning-kmp/blob/62eb5ea0a5a26fd06c69fa499453288a0e2e89dc/src/commonMain/kotlin/fr/acinq/lightning/crypto/LocalKeyManager.kt#L85-L91

> if we want to make it safer we probably need to type this more

Agreed, also the fact that the `KeyManager` has been designed for external signing is a major source of complexity.

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala [hunk 4]
```diff
@@ -38,8 +38,23 @@ object LocalChannelKeyManager {
...
+ * Specifically, there are two paths both of length 8 (256 bits):
+ *   - fundingKeyPath: chosen at random
+ *   - channelKeyPath: sha(fundingPubkey(0))
+ *
+ * The resulting paths looks like so on mainnet:
+ * {{{
+ *  funding:
+ *   - initial/static
+ *     47' / 1' / <fundingKeyPath> / <1 or 0> / 0
+ *   - dynamic
```
[the remark this answers is not in the document]
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala]: Done in 200a4a0cf.

File: eclair-core/src/main/scala/fr/acinq/eclair/json/JsonSerializers.scala (hunk 3 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/json/JsonSerializers.scala]: Removed in 200a4a0cf.

**Comment by pm47** [comment]
Rebased on master.

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala [hunk 5]
```diff
@@ -38,8 +38,25 @@ object LocalChannelKeyManager {
...
+ *   - `fundingKeyPath`: chosen at random using `newFundingKeyPath()`
+ *   - `channelKeyPath`: sha(fundingPubkey(0)) using `ChannelKeyManager.keyPath()`
+ *
+ * The resulting paths looks like so on mainnet:
+ * {{{
+ *  funding:
+ *   - initial funding tx (*):
+ *     47' / 1' / <fundingKeyPath> / <1' or 0'> / 0'
+ *   - splice funding txs (index > 0):
+ *     47' / 1' / <fundingKeyPath> / <1' or 0'> / 6' / <index>
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala:54]: I think we shouldn't add the unnecessary `/ 6' /` and treat splice txs exactly the same as the initial funding tx:

```suggestion
 *     47' / 1' / <fundingKeyPath> / <1' or 0'> / <index'>
```

This way in `override def fundingPublicKey` we don't need to treat the initial funding pubkey differently and can do: `internalKeyPath(channelKeyPath, hardened(fundingTxIndex))`

This guarantees that even if the private key for an old `fundingTxIndex` is leaked, it doesn't leak the private key of the current `fundingTxIndex`, which is nice!

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala [hunk 6]
```diff
@@ -38,8 +38,25 @@ object LocalChannelKeyManager {
...
+ *   (*) The initial funding tx is computed differently for historical reasons.
+ *
+ *  all others (payment, revocation, htlc, etc.):
+ *     47' / 1' / <channelKeyPath> / <1'-5'> / <index>
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala:58]: nit: we should make it obvious that we're hardening the indez:

```suggestion
 *     47' / 1' / <channelKeyPath> / <1'-5'> / <index'>
```

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala (hunk 1 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala]: Simplified in a38095a5f. We are using _sister_ keys, not _child_ keys, so my concern above doesn't apply.

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala (hunk 5 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala]: Done in a38095a5f.

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala (hunk 6 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala:55]: That was wrong actually, there is no index, removed in a38095a5f.

File: eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala [hunk 7]
```diff
@@ -38,8 +38,22 @@ object LocalChannelKeyManager {
...
+ *   - `channelKeyPath`: sha(fundingPubkey(0)) using `ChannelKeyManager.keyPath()`
+ *
+ * The resulting paths looks like so on mainnet:
+ * {{{
+ *  funding:
+ *   - funding txs:
+ *     47' / 1' / <fundingKeyPath> / <1' or 0'> / <index>'
+ *
+ *  others channel basepoint keys (payment, revocation, htlc, etc.):
+ *     47' / 1' / <channelKeyPath> / <1'-5'> / <index>
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala:55]:
```suggestion
 *     47' / 1' / <channelKeyPath> / <1'-5'>
```

Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalChannelKeyManager.scala]: 🤦‍♂️

**Review by t-bast — APPROVED** [review]
:rocket:

**Pull request merged**